api/envoy/extensions/filters/http/waf/v3/waf.proto

```diff
@@ -0,0 +1,223 @@
+    repeated Policy policies = 1[(validate.rules).repeated .min_items = 1];
+  }
+
+  message PolicyGroup {
+    // Name of the Policy Group.
+    string name = 1 [(validate.rules).string.pattern = "^[a-zA-Z0-9]([-a-zA-Z0-9]*[a-zA-Z0-9])?$"];
+    PolicySet policies = 2;
+    RuleMatchCB callback = 3;
+  }
+
+  repeated PolicyGroup policy_groups = 1[(validate.rules).repeated .min_items = 1];
+  Executor executor = 2;
+ }
+
+ // Rule defines a match condition and an associated action.
+ message Policy {
+  string name = 1 (validate.rules).string.pattern = "^[a-zA-Z0-9]([-a-zA-Z0-9]*[a-zA-Z0-9])?$"];
+  Match match = 2;
+  Watcher watcher = 3;
+ }
+
+ // Executor that will execute the actions based on priority or another
+ // custom logic
+
+
+ message Executor {
+   string name = 1;
+   google.protobuf.Any config = 2;
+ }
+
+ // RuleMatchCB that will decide if the rule engine should
+ // 1. Continue evaluating remaining rules that are bound to this callback
+ // 2. Possibly apply request header mutations that Watchers may depend on
+ // 3. Pass along the actions recommended by watchers to others
+
+ message RuleMatchCB {
+   string name = 1;
+   google.protobuf.Any config = 2;
+ }
+
+ // Watcher that inspect the HTTP request and recommend one or more Action
+ message Watcher {
+   string name = 1;
+   google.protobuf.Any config = 2;
+ }
+
+ message Match {
+
+  // Used in the `and_rules` and `or_rules` fields in the `rule` oneof. Depending on the context,
+  // each are applied with the associated behavior.
+  message Set {
+    repeated Match rules = 1 [(validate.rules).repeated .min_items = 1];
+  }
+
+  // Use CidrRangeSet rule instead of large number of CidrRanges combined by or_rule
+  // This rule will evaluate to true, if a given IP is matched aginast any one of the CIDR in this
+  // set
+  message CidrRangeSet {
+    // A list of IP address subnets to be matched against IP address
+    repeated core.v3.CidrRange cidrs = 2 [(validate.rules).repeated .min_items = 1];
+    bool invert_match = 3;
+  }
+
+  oneof rule {
+    option (validate.required) = true;
+
+    // A set of rules that all must match in order to define the action.
+    Set and_rules = 1;
+
+    // A set of rules where at least one must match in order to define the action.
+    Set or_rules = 2;
+
+    // A header (or pseudo-header such as :path or :method) on the incoming HTTP request. Only
+    // available for HTTP request.
+    route.v3.HeaderMatcher header = 3;
+
+    // List of CIDR block that is to be matched against the source IP.
+    CidrRangeSet source_ip = 4;
+  }
+}
+
+message WatcherImpl {
+  message Action {
+    oneof rule {
+      option (validate.required) = true;
+
+      bool nop = 1;
+
+      bool drop_request = 2;
+
+      bool reset_connection = 3;
+
+      bool tarpit_connection = 4;
+
+      string cluster_redirect = 5;
+
+      HttpRedirect http_redirect = 6;
+
+      ExternalService external_service = 6;
+    }
+  }
+
+  message HttpRedirect {
```

Comment:
Please order this before `Action`. Also, should we reuse existing HTTP redirect messages from `RouteConfiguration` here?